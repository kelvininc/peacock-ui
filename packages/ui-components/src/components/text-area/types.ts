import { EventEmitter } from '@stencil/core';
import { EIconName, EOtherIconName } from '../icon/icon.types';

export interface ITextArea {
	/** (optional) Icon to show to the left of the text field */
	icon?: EIconName | EOtherIconName;
	/** (optional) The text to show inside the text area */
	text?: string;
	/** (optional) The placeholder to show in the text area */
	placeholder?: string;
	/** (required) The maximum number of characters allowed */
	maxCharLength: number;
}

export interface ITextAreaEvents {
	/** Emits the current text when there's a change */
	textChange: EventEmitter<string>;
	/** Emits the current text when the text area loses focus */
	textChangeBlur: EventEmitter<string>;
}
