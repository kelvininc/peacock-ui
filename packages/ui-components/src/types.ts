export * from './components';
export * from './utils/types';
export * from './components/action-button/action-button.types';
export * from './components/breadcrumb-item/breadcrumb-item.types';
export * from './components/illustration/illustration.types';
export * from './components/tab-navigation/tab-navigation.types';
export * from './components/text-field/text-field.types';
export * from './components/tooltip/tooltip.types';
export * from './components/summary-card/summary-card.types';
export * from './components/search/search.types';
