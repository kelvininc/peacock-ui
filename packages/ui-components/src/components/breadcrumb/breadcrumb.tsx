import { Component, h, Prop, Host, Event, EventEmitter } from '@stencil/core';
import { IBreadcrumbItem, IBreadcrumbItemEvents } from '../breadcrumb-item/breadcrumb-item.types';

/**
 * @part item - The breadcrumb item.
 */
@Component({
	tag: 'kv-breadcrumb',
	shadow: true
})
export class KvBreadcrumb implements IBreadcrumbItemEvents {
	/** (required) List of breadcrumbs */
	@Prop({ reflect: true }) items: IBreadcrumbItem[] = [];

	/** @inheritdoc */
	@Event() breadcrumbItemClick: EventEmitter<IBreadcrumbItem>;

	render() {
		return (
			<Host>
				<kv-breadcrumb-list>
					{this.items.map(({ active, ...otherProps }, index, array) => (
						<kv-breadcrumb-item active={active ?? index === array.length - 1} part="item" {...otherProps} />
					))}
				</kv-breadcrumb-list>
			</Host>
		);
	}
}
