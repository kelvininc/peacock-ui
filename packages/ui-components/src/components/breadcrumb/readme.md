# *<kv-breadcrumb>*



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<kv-breadcrumb [items]="breadcrumbItems"></kv-breadcrumb>
```


### Javascript

```html
<!-- With items -->
<kv-breadcrumb></kv-breadcrumb>

<script>
	const breadcrumbEl = document.querySelector('kv-breadcrumb');
	breadcrumbEl.items = [
		{
			label: 'Your item label here',
			href: 'Your item link here',
			/**
			 * Check https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target
			 * for a list of possible values. At the moment 'framename' is not supported.
			*/
			target: '_blank',
			/**
			 * Not mandatory, by default the component will set the 'active' attribute to the last item on 
			 * the array, this will disable click events and emphasize the item in bold.
			*/
			active: undefined
		}
	];
</script>
```


### React

```tsx
import React from 'react';
import { KvBreadcrumb } from '@kelvininc/react-ui-components';

export const KvBreadcrumbExample: React.FC = (props) => (
  <>
    {/*-- With items --*/}
    <KvBreadcrumb items={props.items}></KvLink>
  </>
);
```


### Stencil

```tsx
import { Component, h, State } from '@stencil/core';
import { IBreadcrumbItem, EBreadcrumbItemTarget } from 'ui-components';

@Component({
  tag: 'kv-breadcrumb-example',
  styleUrl: 'kv-breadcrumb-example.css',
  shadow: true,
})
export class KvBreadcrumbExample {
	/** 
	 * Depending on your objective you may consider using @Prop or @State
	 * https://stenciljs.com/docs/properties
	 * https://stenciljs.com/docs/state
	 * 
	*/
	@Prop() items: IBreadcrumbItem[] = [
		{
			label: 'Your item label here',
			href: 'Your item link here',
			/**
			 * Check https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target
			 * for a list of possible values. At the moment 'framename' is not supported.
			 *
			 * We also provide a Typescript enum (EBreadcrumbItemTarget) to easily set this value
			*/
			target: '_blank',
			/**
			 * Not mandatory, by default the component will set the 'active' attribute to the last item on 
			 * the array, this will disable click events and emphasize the item in bold.
			*/
			active: undefined
		}
	];

  render() {
    return (
		// With items
		<kv-breadcrumb items={this.items}></kv-breadcrumb>
	);
  }
}
```



## Properties

| Property    | Attribute   | Description                                          | Type                | Default |
| ----------- | ----------- | ---------------------------------------------------- | ------------------- | ------- |
| `items`     | --          | (required) List of breadcrumbs                       | `IBreadcrumbItem[]` | `[]`    |
| `separator` | `separator` | (required) Separator to use between breadcrumb items | `string`            | `'/'`   |


## Events

| Event       | Description                                                 | Type                           |
| ----------- | ----------------------------------------------------------- | ------------------------------ |
| `itemClick` | Emitted when the user clicks on one of the breadcrumb items | `CustomEvent<IBreadcrumbItem>` |


## Dependencies

### Depends on

- [kv-breadcrumb-list](../breadcrumb-list)
- [kv-breadcrumb-item](../breadcrumb-item)

### Graph
```mermaid
graph TD;
  kv-breadcrumb --> kv-breadcrumb-list
  kv-breadcrumb --> kv-breadcrumb-item
  style kv-breadcrumb fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


