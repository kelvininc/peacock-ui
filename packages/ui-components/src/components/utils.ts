import { isNil } from 'lodash-es';
import { DEFAULT_CONFIG } from '../globals/config';
import { UIComponentsConfig } from '../types';

export const getConfig = (): UIComponentsConfig => {
	if (typeof window !== 'undefined') {
		const KvUiComponents = window.KvUiComponents;
		if (!isNil(KvUiComponents) && !isNil(KvUiComponents.config)) {
			return KvUiComponents.config;
		}
	}

	return DEFAULT_CONFIG;
};
