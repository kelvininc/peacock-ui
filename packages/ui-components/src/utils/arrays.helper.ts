export const getArrayOfIndexes = (size: number): number[] =>
	Array(size)
		.fill({})
		.map((_value, index) => index);
