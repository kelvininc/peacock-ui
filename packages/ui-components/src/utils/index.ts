export * from './types';
export * as arraysHelper from './arrays.helper';
export * as clipboardHelper from './clipboard.helper';
export * as cssClassHelper from './css-class.helper';
export * as dateHelper from './date.helper';
export * as searchHelper from './search.helper';
export * as stringHelper from './string.helper';
export * as floatingUIHelper from './floating-ui.helper';
export * as mouseEventHelper from './mouse-event.helper';
