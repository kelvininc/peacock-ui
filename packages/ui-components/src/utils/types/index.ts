export * from './calendars';
export * from './components';
export * from './lib-config';
export * from './intl';
export * from './dates';
