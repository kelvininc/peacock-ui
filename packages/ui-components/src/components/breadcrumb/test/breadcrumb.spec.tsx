import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { cloneDeep } from 'lodash-es';
import { KvBreadcrumb } from '../breadcrumb';
import { BREADCRUMBS_MOCK } from './breadcrumb.mock';
import { EAnchorTarget } from '../../../utils/types';

describe('KvBreadcrumb (unit tests)', () => {
	let page: SpecPage;
	let component: KvBreadcrumb;

	const breadCrumbMock = cloneDeep(BREADCRUMBS_MOCK);

	beforeEach(async () => {
		page = await newSpecPage({
			components: [KvBreadcrumb],
			template: () => <kv-breadcrumb items={breadCrumbMock}></kv-breadcrumb>,
		});
		component = page.rootInstance;
	});

	describe('when the component renders', () => {
		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the correct component items', () => {
			expect(component.items).toEqual([
				{
					label: 'Homepage',
					href: 'https://kelvin.ai',
					target: EAnchorTarget.NewTab
				},
				{
					label: 'Product',
					href: 'https://kelvin.ai/product/',
					target: EAnchorTarget.NewTab
				},
				{
					label: 'Contact Form',
					href: 'https://kelvin.ai/product#form',
					target: EAnchorTarget.NewTab
				}
			]);
		})
	});
});
