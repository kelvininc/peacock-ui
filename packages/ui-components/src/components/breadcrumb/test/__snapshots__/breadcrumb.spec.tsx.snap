// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KvBreadcrumb (unit tests) when the component renders should match the snapshot 1`] = `
<kv-breadcrumb separator="/">
  <mock:shadow-root>
    <kv-breadcrumb-list separator="/">
      <kv-breadcrumb-item href="https://kelvin.ai" label="Homepage" target="_blank"></kv-breadcrumb-item>
      <kv-breadcrumb-item href="https://kelvin.ai/product/" label="Product" target="_blank"></kv-breadcrumb-item>
      <kv-breadcrumb-item active="" href="https://kelvin.ai/product#form" label="Contact Form" target="_blank"></kv-breadcrumb-item>
    </kv-breadcrumb-list>
  </mock:shadow-root>
</kv-breadcrumb>
`;
